### ISSUE TITLE
[BUG] "A mandatory record is missing" when calling `delete_link` just after a `get_links` from a node with an empty gossip arc clamping.

### ISSUE THREAD

**Issue by guillemcordoba** [holochain/holochain#4012]
**Describe the bug**

An error shows up when trying to call `delete_link` from a node with empty gossip arc clamping that just called `get_links` for those links.

**Expected behavior**

1) Calling `delete_link` succeeds on nodes with empty arc clamping.
2) Calling `delete_link` succeeds on nodes with empty arc clamping independently on whether you have called `get_links` previously or not (you may have got the create link hash from somewhere else and are trying to delete it directly)

I think these two concerns would be addressed if [this call](https://github.com/holochain/holochain/blob/649ad4c97ecb0bb113139d1aff688b689853b292/crates/holochain/src/core/ribosome/host_fn/delete_link.rs#L42) goes to the network instead of just being local.

**System information:**
 - OS: [e.g. iOS]: NixOS
 - Holochain and HDK Version: holochain repo, zero-arc-validation-bug branch

**Steps to reproduce**

Here is the PR that adds the test that can reproduce this issue: https://github.com/holochain/holochain/compare/zero-arc-validation-bug...zero-arc-validation-bug-delete-link?expand=1

**Comment by jost-s** [comment]
Confirmed that is happening, independent of arc clamping. If a link is deleted before the create action has made it to the deleter _and_ the deleter is an authority, the create action will not be fetched from the network.

**Comment by jost-s** [comment]
My proposal to resolve this bug is to
1. Use `GetOptions::network()` instead of [`GetOptions::local()`](https://github.com/holochain/holochain/blob/develop/crates/holochain/src/core/ribosome/host_fn/delete_link.rs#L42). This is for sure a bug, it is wrong to restrict the dht get to local databases.
2. Enable authorities to fetch missing actions [where they're currently not fetching anything they're supposedly an authority for](https://github.com/holochain/holochain/blob/develop/crates/holochain_cascade/src/lib.rs#L782-L783).

**Comment by jost-s** [comment]
@ddd-mtl Which problem are you encountering? You call `get_links` and get the expected links, but to delete a link you need to wait first?

**Comment by ddd-mtl** [comment]
Right after I get a "new" link after calling `get_links()`, I call a zome function to do a `delete_link()` on it. If I do it immediately I get the "A mandatory record is missing" error. If I wait a second before calling that zome function, it works.

**Comment by jost-s** [comment]
Then please provide a link to the app where you're having that problem, because I'm unable to reproduce it without an empty arc setting or without calling `get_links` first. In my test, when the link is received, it can immediately be deleted.

**Comment by ddd-mtl** [comment]
https://github.com/lightningrodlabs/vines/blob/28a665f813c43e8ad1b5fef8231d42e324081eb7/webapp/src/vines-page.ts#L796
In my happ, the agent creating the link does remote_signal() to the other peer to tell them to call get_links(). Have you tried that?

**Comment by jost-s** [comment]
No, I haven't, but I see now how the problem comes to be in that scenario. The other peer doesn't have the new link yet, so when it calls `get_links` it does get it into their cache database, not the DHT database. When trying to delete the link, the original create is looked up, but the query to look it up fails as the required fields in the cache database are not populated. That's why it fails.

That's a bug, and that's the same reason why deleting a link fails in the empty arc scenario. My previous proposal is not appropriate for this actually. What we've thought of doing is populating the fields in the cache database, so that the lookup will succeed. That should be the least invasive fix. I'll see if I can adjust the reproduction and then submit the PR for the fix.

**Comment by ddd-mtl** [comment]
As a happ dev I expect anything that get_links returns to be available, so
maybe there is a solution where get_links doesn't returns links that are
not actually fully cached / available?

**Comment by jost-s** [comment]
I've spent the past two days trying to write a good test for this and fixing it and the issue has transformed under my feet. There are three databases for every DNA for every agent in Holochain, `authored`, `dht` and `cache`. When doing `get_links`, returned links are written to the cache database. I thought the issue was that some fields are not populated which then a lookup of the `CreateLink` are depended upon when deleting a link. But that was not the issue.

The problem is instead that the lookup query of the `CreateLink` depends on a certain type of DHT op. `get_links` only fetches the op type `RegisterAddLink`, but the query looks for `StoreRecord`. `StoreRecord` comes in with publish or gossip when peers are the record authority of the link. That's why waiting 1 second resolves the problem in your case @ddd-mtl.

It doesn't help with @guillemcordoba's problem, however, and would also happen in a sharded network, where the peer that wants to delete a link is not the record authority.

Unfortunately this issue is not easy to resolve and tests would have to be written first to make sure that existing functionality is not broken. Which makes this a larger task to complete.

I'm sorry that I don't have a solution for this problem at the moment, but I need to go back to working on the kitsune integration for now. Waiting until the relevant ops have made it to the peer is the workaround for the time being in an unsharded network, and we're aiming to fix the sharded scenario during the workflow review.

**Comment by jost-s** [comment]
In short the problem is:
- `get_links` fetches all create link actions as `RegisterAgentActivity` ops
- `delete_link` looks for the action that created the link as a `StoreRecord` op

So while publish or gossip haven't gotten in the `StoreRecord` op corresponding to the link to be deleted, which does never happen on a zero arc node, the link can not be deleted.

**Comment by artbrock** [comment]
Hmmm... So... Let's think this through a minute... Was there any reason that the delete_link action was getting the `StoreRecord` op in the first place?

My guess is that this was some kind of confusion that emerged from the SQL database conversion switching to a more op-centric pattern when the DHT was previously centered on action & entry hashes (which is what get() commands are typically getting ... not ops.) So the logic was probably that we need the record before deleting the record.

But `RegisterAddLink` carries the whole record, doesn't it? (both the header & entry)  So... the content of the data is the same. I think the only question is the possibility of a difference in validation status on the op or if the record authority would have some different meta-data about the record (like previous delete actions). I don't think its likely the record authority will need to have stricter link validation than the basis hash holder serving out the links. So we can probably write off a validation difference. And I can't think of any meta-data the record authority should receive that the basis hash authority wouldn't, so I don't see any need to be using the `StoreRecord` op rather than `RegisterAddLink`.  It was probably just not thought through that we'd likely have different ops cached than that one.

So... yes... I support going with the third solution in the list:

> delete_links looks up the RegisterAddLink op instead of the StoreRecord op (requires a new dht_get_* call)

**Comment by ThetaSinner** [comment]
Suggested from refinement:
- Pass the create action for the link to the delete
- Delete link then figures out the structure of the delete for you

**Comment by jost-s** [comment]
resolved by #4945
